File: modules/widgets/lib/rin/widgets/content/ImageWidget.cpp
#include "rin/widgets/content/ImageWidget.hpp"
#include "rin/graphics/DeviceImage.hpp"
#include "rin/graphics/ResourceManager.hpp"
#include "rin/widgets/graphics/SimpleBatchedDrawCommand.hpp"
#include "rin/widgets/graphics/WidgetDrawCommands.hpp"

void ImageWidget::SetTextureId(const int& textureId)
{
    auto isNew = _textureId != textureId;
    _textureId = textureId;
    if (isNew)
    {
        CheckSize();
    }
}

int ImageWidget::GetTextureId() const
{
    return _textureId;
}

void ImageWidget::SetBorderRadius(const Vec4<float>& borderRadius)
{
    _borderRadius = borderRadius;
}

Vec4<float> ImageWidget::GetBorderRadius() const
{
    return _borderRadius;
}

void ImageWidget::SetTint(const Vec4<float>& tint)
{
    _tint = tint;
}

Vec4<float> ImageWidget::GetTint() const
{
    return _tint;
}

void ImageWidget::CollectContent(const TransformInfo& transform, WidgetDrawCommands& drawCommands)
{
    if (_textureId < 0)
    {
        drawCommands.Add(
            SimpleBatchedDrawCommand::Builder{}
            .AddRect(
                GetContentSize(),
                transform.transform,
                GetBorderRadius(),
                GetTint()
            )
            .Finish()
        );
    }
    else
    {
        drawCommands.Add(
            SimpleBatchedDrawCommand::Builder{}
            .AddTexture(
                _textureId,
                GetContentSize(),
                transform.transform,
                GetBorderRadius(),
                GetTint()
            )
            .Finish()
        );
    }
}

Vec2<float> ImageWidget::ComputeContentSize()
{
    if (auto vkImage = GraphicsModule::Get()->GetResourceManager()->GetTextureImage(GetTextureId()))
    {
        auto extent = vkImage->GetExtent();
        return {static_cast<float>(extent.width), static_cast<float>(extent.height)};
    }
    return {0.0f, 0.0f};
}

File: rin.Widgets/Content/ImageWidget.cs
using System.Runtime.InteropServices;
using rin.Graphics;
using rin.Core.Math;
using rin.Widgets.Graphics;
using rin.Widgets.Graphics.Quads;

namespace rin.Widgets.Content;

/// <summary>
///     Draw's a <see cref="Texture" /> if provided or a colored rectangle. Supports tint.
/// </summary>
public class ImageWidget : Widget
{

    private int _textureId = -1;

    public ImageWidget()
    {
    }

    public int TextureId
    {
        get => _textureId;
        set
        {
            _textureId = value;
            TryUpdateDesiredSize();
        }
    }

    public Color Tint { get; set; } = new Color(1.0f);

    public Vector4<float> BorderRadius { get; set; } = new(0.0f);

    protected override void OnDispose(bool isManual)
    {
        base.OnDispose(isManual);
    }
    
    protected override Vector2<float> ComputeDesiredContentSize()
    {
        if (SGraphicsModule.Get().GetResourceManager()
                .GetTextureImage(TextureId) is { } texture)
        {
            return new Vector2<float>
            {
                X = texture.Extent.width,
                Y = texture.Extent.height
            };
        }

        return new Vector2<float>();
    }
    
    protected override void OnRemovedFromSurface(Surface surface)
    {
        base.OnRemovedFromSurface(surface);
    }

    public override void CollectContent(TransformInfo info, DrawCommands drawCommands)
    {
        //throw new NotImplementedException();
        if (TextureId != -1)
        {
            drawCommands.Add(new QuadDrawCommand([new Quad(GetContentSize(), info.Transform)
            {
                TextureId = TextureId,
                Color = Tint,
                BorderRadius = BorderRadius
            }]));
        }
    }
}